lib/middleman-prismic/commands/prismic.rb

```diff
@@ -29,40 +23,68 @@ def self.exit_on_failure?
       end
 
       def prismic
-        ::Middleman::Application.server.inst
-        reference = MiddlemanPrismic.options.release
-
-        Dir.mkdir('data') unless File.exists?('data')
-
-        FileUtils.rm_rf(Dir.glob('data/prismic_*'))
+        create_directories
+        output_available_documents
+        output_references
+        output_custom_queries
+      end
 
-        api = ::Prismic.api(MiddlemanPrismic.options.api_url)
-        response = api.form('everything').submit(api.ref(reference))
+      private
 
-        available_documents = []
-        response.each {|d| available_documents << d.type}
+      def create_directories
+        if File.exists?("#{DATA_DIR}/prismic")
+          FileUtils.rm_rf(Dir.glob("#{DATA_DIR}/prismic"))
+        end
 
-        available_documents.uniq!
+        Dir.mkdir("#{DATA_DIR}") unless File.exists?("#{DATA_DIR}")
+        Dir.mkdir("#{DATA_DIR}/prismic")
+      end
 
-        available_documents.each do |document_type|
-          documents = response.select{|d| d.type == document_type}
-          File.open("data/prismic_#{document_type.pluralize}", 'w') do |f|
-            f.write(Hash[[*documents.map.with_index]].invert.to_yaml)
-          end
+      def output_available_documents
+        api_response.map(&:type).uniq.each do |document_type|
+          document_dir = "#{DATA_DIR}/prismic/#{document_type.pluralize}"
+          documents = api_response.select { |d| d.type == document_type }
+          write_collection(document_dir, documents)
         end
+      end
 
-        File.open('data/prismic_reference', 'w') do |f|
+      def output_references
+        File.open("#{DATA_DIR}/prismic/reference.yml", "w") do |f|
           f.write(api.master_ref.to_yaml)
         end
+      end
+
+      def output_custom_queries
+        Middleman::Prismic.options.custom_queries.each do |key, value|
+          document_dir = "#{DATA_DIR}/prismic/custom_#{key}"
+          response = api.form("everything").query(*value).submit(api.master_ref)
+          write_collection(document_dir, response)
+        end
+      end
+
+      def api_response
+        @api_response ||= api.form('everything').submit(api_reference)
+      end
+
+      def api_reference
+        api.ref(Middleman::Prismic.options.release)
+      end
+
+      def api
+        @api ||= ::Prismic.api(Middleman::Prismic.options.api_url)
+      end
+
+      def write_collection(dir, collection)
+        Dir.mkdir(dir) unless File.exists?(dir)
```

Comment:
Maybe a race condition here?

Reply:
@georgebrock I'm not sure I understand why. Is it because something else could create the directory between the check and create?

Reply:
@danbee Yes, or something could remove the directory after the check before it's used. It's unlikely, but it's generally preferable to use exception handling instead of conditionals for file system operations, e.g.

```ruby
begin
  Dir.mkdir(dir)
rescue Errno::EEXIST
end
```

`FileUtils.mkdir_p` might also be safer.